### PR TITLE
[libc] move __stack_chk_fail to src/ from startup/

### PR DESCRIPTION
**Pull request by nickdesaulniers**
__stack_chk_fail should be provided by libc.a, not startup files.

Add __stack_chk_fail to existing linux and arm entrypoints. On Windows (when
not targeting MinGW), it seems that the corresponding function identifier is
__security_check_cookie, so no entrypoint is added for Windows. Baremetal
targets also ought to be compileable with `-fstack-protector*`

There is no common header for this prototype, since calls to __stack_chk_fail
are meant to be inserted by the compiler upon function return when compiled
`-fstack-protector*`.

### CODE REVIEW

File: libc/src/compiler/generic/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+add_entrypoint_object(
+  __stack_chk_fail
+  SRCS
+    __stack_chk_fail.cpp
+  HDRS
+    ../__stack_chk_fail.h
```
Comment (michaelrj-google) [inline libc/src/compiler/generic/CMakeLists.txt:6]: it looks like this header didn't get included in your patch?

Comment (nickdesaulniers) [reply libc/src/compiler/generic/CMakeLists.txt:6]: fixed in https://github.com/llvm/llvm-project/pull/75863/commits/7c26cda7a7681ec11e2ccc4c52f3dfb1f6685e98

**Comment by nickdesaulniers** [comment]
I talked with @michaelrj-google about adding a  `EXPECT_DEATH` for this, too.

**Comment by SchrodingerZhu** [comment]
As a side note, I was to move `__stack_chk_fail` to a separate file once inside startup once https://github.com/llvm/llvm-project/pull/75413 merged. But it does make more sense to move it to `src`.

**Review by michaelrj-google — APPROVED** [review]
LGTM, though if possible it would be nice to add a test to check the actual stack check mechanism (i.e. overwriting the stack cookie intentionally to cause a crash)

**Comment by nickdesaulniers** [comment]
> LGTM, though if possible it would be nice to add a test to check the actual stack check mechanism (i.e. overwriting the stack cookie intentionally to cause a crash)

@michaelrj-google done in https://github.com/llvm/llvm-project/pull/75863/commits/f5e86b34ff5b100114b98f9a60137e06fbc013d5

**Review by Prabhuk — APPROVED** [review]
LGTM

File: libc/test/src/compiler/stack_chk_guard_test.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
+//===-- Unittests for assert ----------------------------------------------===//
```
Comment (michaelrj-google) [inline libc/test/src/compiler/stack_chk_guard_test.cpp:1]: nit: fix header

**Review by michaelrj-google — APPROVED** [review]
LGTM

File: libc/test/src/compiler/stack_chk_guard_test.cpp (hunk 2 above)
Comment (nickdesaulniers) [reply libc/test/src/compiler/stack_chk_guard_test.cpp]: done in 13df4945c04e4b145fd0e60e25020c2743537812

**Comment by nickdesaulniers** [comment]
follow up fix for arm: https://github.com/llvm/llvm-project/pull/75962

**Pull request merged**